### ISSUE TITLE
Extended support for boolean expressions.

### ISSUE THREAD

**Issue by RobertZenz** [uklimaschewski/EvalEx#128]
I'm using EvalEx for [a simple calculator application](https://gitlab.com/RobertZenz/jMathPaper/) and during [the first round of feedback](https://www.reddit.com/r/codeprojects/comments/6xop8c/jmathpaper_a_simple_calculator_working_similar_to/) some sort of support for boolean operators was requested.

To be exact support for bitwise operators:

 * and
 * or
 * xor
 * maybe left shift (<<)
 * maybe right shift (>>)

So that expressions like `true or false` can be evaluated. But there seem to be a lot of edge cases (like `5<<0.4`) which might make this tricky.

Additionally it would be interesting to have the possibility to see if an expression that has been evaluated evaluates to a boolean value (`true` or `false`).

**Comment by uklimaschewski** [comment]
EvalEx has the ability to work with boolean expressions, see the README. Any value other than zero is treated as true. Additional bitwise operators could be added without any problem.

**Comment by RobertZenz** [comment]
Let me rephrase the last part, please. It would be interesting to have the ability to see whether an evaluated result of an Expression is a boolean or not.

    Expression expression = new Expression("1 > 0");
    BigDecimal result = expression.eval();
    
    System.out.println("Expression returned boolean: " + expression.wasBoolean());

Obviously, I'm kinda torn how to implement such a functionality, because the example above isn't the best API for that.

**Comment by Sxtanna** [comment]
I agree, this would be a great addition. @RobertZenz 

You could possibly wrap the result in a class that extends BigDecimal.
(as to not break functionality, + backwards compat) @uklimaschewski 

And in this class have possibly several methods for checking what kind of result you've received.

**Comment by uklimaschewski** [comment]
The problem is that you can mix arithmetic and boolean operators/functions in an expression.
For example, `x > a + b`.
here, we could say that the last operator/function evaluated is boolean (`x >`) and therefore, the expression result is also a boolean.
If there was an `isBoolean` flag in each function/operator, then we could determine if the last operator/function evaluated was boolean or not and also flag the expression as boolean or not.

**Comment by Sxtanna** [comment]
Is that actually a problem? It seems that would allow this to be implemented fairly easily? @uklimaschewski

**Comment by uklimaschewski** [comment]
No, no problem. Thinking about it, implementing an isBooleanExpression() function would even be possible without evaluating. You just have to parse the expression and get the RPN through shunting yard. 
see the following function:
https://github.com/uklimaschewski/EvalEx/blob/8bc12f99c8c330028ae3e3917fc70e37a8c91be8/src/com/udojava/evalex/Expression.java#L1311
Look at the last token in the returned RPN list, then you have it.

**Comment by uklimaschewski** [comment]
@RobertZenz I will try to find some time this weekend to give it shot, is that ok?

**Comment by RobertZenz** [comment]
@Sxtanna Extending BigDecimal is an interesting idea, I thought it was `final`, but isn't.

@uklimaschewski Absolutely, thank you very much for the fast response.

**Comment by uklimaschewski** [comment]
I have added the isBoolean() method to the Expression class in commit a4cb21f6bf26419027ef158a29a3c8f43a214834.
Looks good, works exactly like I explained. The IF function is handled special. Here I take it that the IF is boolean, if the third parameter is boolean.

**Comment by uklimaschewski** [comment]
@RobertZenz Please give a short feedback, when you had time to validate. Then we can close this issue.

**Comment by RobertZenz** [comment]
Yepp, works awesomely. Thank you.
